events/ratelimiting/coalescing.go

```diff
@@ -0,0 +1,256 @@
+	initialDelay := time.Millisecond * 500
+	if opts.InitialDelay != nil {
+		initialDelay = *opts.InitialDelay
+	}
+	if initialDelay < 0 {
+		return nil, errors.New("initial delay must be > 0")
+	}
+
+	maxDelay := time.Second * 5
+	if opts.MaxDelay != nil {
+		maxDelay = *opts.MaxDelay
+	}
+	if maxDelay < 0 {
+		return nil, errors.New("max delay must be > 0")
+	}
+
+	if maxDelay < initialDelay {
+		return nil, errors.New("max delay must be >= base delay")
+	}
+
+	if opts.MaxPendingEvents != nil && *opts.MaxPendingEvents <= 0 {
+		return nil, errors.New("max pending events must be > 0")
+	}
+
+	return &coalescing{
+		initialDelay:     initialDelay,
+		maxDelay:         maxDelay,
+		maxPendingEvents: opts.MaxPendingEvents,
+		currentDur:       initialDelay,
+		backoffFactor:    1,
+		inputCh:          make(chan struct{}),
+		closeCh:          make(chan struct{}),
+		clock:            clock.RealClock{},
+	}, nil
+}
+
+// Run runs the rate limiter. It will begin rate limiting events after the
+// first event is received.
+func (c *coalescing) Run(ctx context.Context, ch chan<- struct{}) error {
+	if !c.running.CompareAndSwap(false, true) {
+		return errors.New("already running")
+	}
+
+	// Prevent wg race condition on Close and Run.
+	c.lock.Lock()
+	c.wg.Add(1)
+	c.lock.Unlock()
+	defer c.wg.Done()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		// If the timer doesn't exist yet, we're waiting for the first event (which
+		// will fire immediately when received).
+		if !c.hasTimer.Load() {
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-c.closeCh:
+				cancel()
+				return nil
+
+			case <-c.inputCh:
+				c.handleInputCh(ctx, ch)
+			}
+		} else {
+			// We already have a timer running, so we're waiting for either the timer
+			// to fire, or for a new event to arrive.
+			select {
+			case <-ctx.Done():
+				return nil
+			case <-c.closeCh:
+				cancel()
+				return nil
+
+			case <-c.inputCh:
+				c.handleInputCh(ctx, ch)
+			case <-c.timer.C():
+				c.handleTimerFired(ctx, ch)
+			}
+		}
+	}
+}
+
+func (c *coalescing) handleInputCh(ctx context.Context, ch chan<- struct{}) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	switch {
+	case !c.hasTimer.Load():
+		// We don't have a timer yet, so this is the first event that has fired. We
+		// fire the event immediately, and set the timer to fire again after the
+		// initial delay.
+		c.timer = c.clock.NewTimer(c.initialDelay)
+		c.hasTimer.Store(true)
+		c.fireEvent(ctx, ch)
+
+	default:
+		// If maxPendingEvents is set and we have reached it then fire the event
+		// immediately.
+		if c.maxPendingEvents != nil && c.pendingEvents >= *c.maxPendingEvents {
+			c.fireEvent(ctx, ch)
+			return
+		}
+
+		if !c.timer.Stop() {
+			<-c.timer.C()
+		}
+
+		// Setup backoff. Backoff is exponential. If initial is 500ms and max is
+		// 5s, the backoff will follow:
+		// 500ms, 1s, 2s, 4s, 5s, 5s, 5s, ...
+		if c.currentDur < c.maxDelay {
+			c.backoffFactor *= 2
```

Comment:
Do we need some jitter?

Reply:
I initially did have some jitter but I don't think it makes sense in this scenario since we are not concerned with contention as we are coalescing the requests and not responding to clients which would all re-request at the same time on backoff.